### PR TITLE
Add stored procedure for enumerating geckoview versions in fenix nightly

### PR DESCRIPTION
**Pull request by acmiyaguchi**
See #1402 

This was tested in the `glam-fenix-dev` project using the following SQL:

<details>
<summary>Click to expand</summary>

```sql
DECLARE
  results ARRAY<STRUCT<build_hour DATETIME,
  geckoview_version STRING,
  n_builds INT64>>;
CREATE OR REPLACE PROCEDURE
  analysis.fenix_nightly_geckoview_versions(
    IN reference_date DATE,
    IN history_days INT64,
    OUT build_hour_version ARRAY<
      STRUCT<build_hour DATETIME, geckoview_version STRING, n_builds INT64>
    >
  )
BEGIN
  -- https://github.com/mozilla/bigquery-etl/blob/56defb1061445115277d90def06ad41185e519b5/sql/moz-fx-data-shared-prod/udf/fenix_build_to_datetime/metadata.yaml#L1-L23
  SET build_hour_version = (
    WITH extracted AS (
    -- We'll look at the metrics ping to estimate the major geckoview version
    -- the metrics section is aliased, so we must rename the table for this to
    -- work appropriately
      SELECT
        submission_timestamp,
        client_info.app_build,
        metrics.string.geckoview_version,
      FROM
        `moz-fx-data-shared-prod`.org_mozilla_fenix.metrics AS t1
      WHERE
        mozfun.norm.fenix_app_info('org_mozilla_fenix', client_info.app_build).channel = 'nightly'
      UNION ALL
      SELECT
        submission_timestamp,
        client_info.app_build,
        metrics.string.geckoview_version,
      FROM
        `moz-fx-data-shared-prod`.org_mozilla_fenix_nightly.metrics AS t1
      UNION ALL
      SELECT
        submission_timestamp,
        client_info.app_build,
        metrics.string.geckoview_version,
      FROM
        `moz-fx-data-shared-prod`.org_mozilla_fennec_aurora.metrics AS t1
    ),
    transformed AS (
      SELECT
        app_build,
        geckoview_version,
        datetime_TRUNC(
          `moz-fx-data-shared-prod`.udf.fenix_build_to_datetime(app_build),
          HOUR
        ) AS build_hour
      FROM
        extracted
      WHERE
        DATE(submission_timestamp) >= DATE_SUB(reference_date, INTERVAL history_days DAY)
    ),
    grouped_build_hours AS (-- count the number of geckoview versions for each build hour row over an interval
      SELECT
        build_hour,
        geckoview_version,
        COUNT(*) AS n_builds
      FROM
        transformed 
      WHERE
        geckoview_version IS NOT NULL
        AND app_build IS NOT NULL
        AND build_hour IS NOT NULL
      GROUP BY
        build_hour,
        geckoview_version
      HAVING
        n_builds > 5
      ORDER BY
        build_hour DESC,
        geckoview_version
    ),
    top_build_hours AS (-- get the geckoview version for the build hour that has the most number of rows
      SELECT
        ROW.*
      FROM
        (
          SELECT
            ARRAY_AGG(t ORDER BY t.n_builds DESC LIMIT 1)[OFFSET(0)] ROW
          FROM
            grouped_build_hours t
          GROUP BY
            build_hour
        )
    ),
    enumerated_build_hours AS (-- enumerate all of the build hours that we care about. We have a small margin that we'll
      -- use so we fill in null values for the rolling average of number of rows
      SELECT
        datetime(TIMESTAMP) AS build_hour
      FROM
        UNNEST(
          GENERATE_TIMESTAMP_ARRAY(
            TIMESTAMP_SUB(
              TIMESTAMP_TRUNC(CAST(reference_date AS timestamp), HOUR),
              INTERVAL history_days + 2 DAY
            ),
            TIMESTAMP_TRUNC(CAST(reference_date AS timestamp), HOUR),
            INTERVAL 1 HOUR
          )
        ) AS TIMESTAMP
    ),
    estimated_version AS (
      -- right join
      SELECT
        build_hour,
        MAX(geckoview_version) OVER (
          ORDER BY
            build_hour ASC
          ROWS BETWEEN
            UNBOUNDED PRECEDING
            AND CURRENT ROW
        ) AS geckoview_version,
        AVG(n_builds) OVER (
          ORDER BY
            build_hour ASC
          ROWS BETWEEN
            48 PRECEDING
            AND CURRENT ROW
        ) AS n_builds
      FROM
        top_build_hours
      RIGHT JOIN
        enumerated_build_hours
      USING
        (build_hour)
    )
    SELECT
      ARRAY_AGG(
        STRUCT<build_hour DATETIME, geckoview_version STRING, n_builds INT64>(
          build_hour,
          geckoview_version,
          CAST(n_builds AS INT64)
        )
        ORDER BY
          build_hour
      )
    FROM
      estimated_version
    WHERE
      build_hour >= DATE_SUB(reference_date, INTERVAL history_days DAY)
  );
END;

CALL analysis.fenix_nightly_geckoview_versions(DATE "2020-10-01", 28, results);

SELECT
  *
FROM
  UNNEST(results)
  order by build_hour desc
```
</details>

Removing the stored procedure bits makes this easier to test in STMO. See the following query: https://sql.telemetry.mozilla.org/queries/75453/source#188356

image

### CODE REVIEW

**Comment by acmiyaguchi** [comment]
It looks like the tests might be replacing the reference to `mozfun.norm.fenix_app_info`, despite it being in a sibling dataset.

https://github.com/mozilla/bigquery-etl/blob/0d51459bd1bce3ac74b72bb28aae3fcbf4f4da0e/bigquery_etl/routine/parse_routine.py#L275-L294

**Comment by fbertsch** [comment]
> It looks like the tests might be replacing the reference to `mozfun.norm.fenix_app_info`, despite it being in a sibling dataset.

That behavior is expected, but the function is now defined at the top of the test. Does removing the backticks from the function call fix the error?

**Comment by acmiyaguchi** [comment]
It's a scoping issue in particular. Since the UDF/stored procedures are deployed to temporary dataset, any referenced udfs are inlined in the global scope. Either referencing the mozfun function in production or putting a reference to the udf in the local scope of the procedure would allow this to work.

**Comment by acmiyaguchi** [comment]
@fbertsch this is ready for review with the new patches made. It turns out that the store procedure needs to be ignored anyways in the tests because of permission issues.

File: sql/mozfun/glam/fenix_nightly_geckoview_versions/stored_procedure.sql [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        geckoview_version,
+        -- Truncate to the hour, since older builds give minute resolution.
+        datetime_TRUNC(
+          `moz-fx-data-shared-prod`.udf.fenix_build_to_datetime(app_build),
+          HOUR
+        ) AS build_hour
+      FROM
+        extracted
+      WHERE
+        DATE(submission_timestamp) >= DATE_SUB(reference_date, INTERVAL history_days DAY)
```
Comment (fbertsch) [inline sql/mozfun/glam/fenix_nightly_geckoview_versions/stored_procedure.sql:72]: Should this be `BETWEEN DATE_SUB(reference_date, INTERVAL history_days DAY) AND reference_date`?

**Review by fbertsch — COMMENTED** [review]
Should this be a derived table that calculates this every day?

File: sql/mozfun/glam/fenix_nightly_geckoview_versions/stored_procedure.sql (hunk 1 above)
Comment (acmiyaguchi) [reply sql/mozfun/glam/fenix_nightly_geckoview_versions/stored_procedure.sql:72]: Yes, that's a good catch.

**Review by acmiyaguchi — COMMENTED** [review]
Yeah it should, I'm going to be building a derived table using this procedure as a basis.

**Comment by acmiyaguchi** [comment]
Closed in favor of #1419. There are difficulties using stored procedures in incremental queries, unfortunately.

**Pull request closed without merging**